prov/verbs/src/verbs_msg_ep.c

```diff
@@ -561,30 +561,37 @@ static inline int fi_ibv_poll_reap_unsig_cq(struct fi_ibv_msg_ep *ep)
 	return FI_SUCCESS;
 }
 
+static inline ssize_t
+fi_ibv_unsignaled_send(struct fi_ibv_msg_ep *ep, struct ibv_send_wr *wr)
+{
+	int ret = FI_IBV_INVOKE_POST(send, send, ep->id->qp, wr, NULL);
+	if (OFI_UNLIKELY(ret))
+		return ret;
+
+	/* `+ 1` means to make space for the next unsignaled send. */
+	if ((ofi_atomic_inc32(&ep->scq->sends_outstanding) + 1) >=
+					ep->info->tx_attr->size) {
+		ret = fi_ibv_poll_reap_unsig_cq(ep);
+		if (OFI_UNLIKELY(ret))
+			return -FI_EAGAIN;
+	}
+	return ret;
+}
+
 /* WR must be filled out by now except for context */
 static inline ssize_t
 fi_ibv_send(struct fi_ibv_msg_ep *ep, struct ibv_send_wr *wr, void *context)
 {
-	struct fi_ibv_wre *wre = NULL;
-	int ret;
-
 	if (wr->send_flags & IBV_SEND_SIGNALED) {
-		ret = fi_ibv_prepare_signal_send(ep, wr, &wre, context);
+		struct fi_ibv_wre *wre;
+		int ret = fi_ibv_prepare_signal_send(ep, wr, &wre, context);
 		if (OFI_UNLIKELY(ret))
 			return ret;
+		return FI_IBV_INVOKE_POST(send, send, ep->id->qp, wr,
+					  FI_IBV_RELEASE_WRE(ep, wre));
 	} else {
-		if (ofi_atomic_inc32(&ep->scq->sends_outstanding) >=
-						ep->info->tx_attr->size) {
-			ret = fi_ibv_poll_reap_unsig_cq(ep);
-			if (OFI_UNLIKELY(ret)) {
-				ofi_atomic_dec32(&ep->scq->sends_outstanding);
-				return -FI_EAGAIN;
-			}
-		}
+		return fi_ibv_unsignaled_send(ep, wr);
```

Comment:
Wasn't there a recent change to always signal all send WRs?

Reply:
Yes. I think the function should be renamed to reflect the intent correctly. Something better than "fi_ibv_send_read_cq_if_needed".